Scrooge codegen improvements

kevinoliver opened the pull request:
### Problem

Scrooge codegen does not do fingerprinting for enough facets and is
missing support for include_path and default-java-namespace.

### Solution

Add support for include-path and default-java-namespace.

Fix the fingerprinting to include namespace_map, include_path and
default_java_namespace.

### Result

Scrooge support is now good enough for to be used for development
of Scrooge itself.

stuhood commented on src/python/pants/backend/codegen/thrift/java/thrift_defaults.py:
```diff
@@ -25,12 +25,21 @@ def register_options(cls, register):
+             help='The default Java namespace to generate for java_thrift_library targets.')
+    register('--namespace-map', type=dict, advanced=True, default={},
+             help='The default namespace map to generate for java_thrift_library targets, {old: new}.')
+    register('--include-paths', type=list, advanced=True, default=[],
```
Can you add a bit more information about how this would be used? We would generally prefer that people explicitly declare the dependencies of a target, which provides enough information to infer the include paths automatically.

So on the surface, I don't think this option makes sense: it would bypass change detection and allow a target to depend on an entire other subdirectory without declaring it as a dependency. In essence, it would exacerbate the internal ticket we have about this: `DPB-5639`, which states:
```
Currently, scrooge gen adds the source root of targets to the import paths of the global scrooge gen. This means that a target that depends on a thrift file [under a particular source root] can include any thrift [under that source root].

When there is an include in a thrift file whose target doesn't have matching dependency listings, incompatible changes to the included file can pass SQ, but will cause subsequent builds to fail.
```

kevinoliver replied:
@stuhood what do you think about omitting these from defaults entirely? i just added these as cargo-cult.

kevinoliver commented:
reviewers: apologies in advance for my lack of python and pants knowledge. 

also note that i tested my changes on a earlier commit and it looks like some of the codegen code got moved around since then. that means my confidence level is not super high given the levels of test coverage i experienced.

stuhood commented:
@kevinoliver : It should be relatively easy to add an integration test for this by cloning https://github.com/pantsbuild/pants/blob/master/contrib/scrooge/tests/python/pants_test/contrib/scrooge/tasks/test_thrift_linter_integration.py into a new file `test_scrooge_integration.py`, and then invoking the `compile` or `gen` task there.

There is also an `@ensure_cached` annotation that you can add to your test ([example](https://github.com/pantsbuild/pants/blob/master/tests/python/pants_test/backend/jvm/tasks/test_checkstyle_integration.py#L126)) to validate that some number of distinct artifacts are written to the cache.

kevinoliver commented:
Thanks for the review. I've pushed a new version:
* Removed the defaults for include_paths
* Added integration tests for scrooge_gen

I tried, unsuccessfully, to add the caching test. Hopefully that isn't a blocker.

Let me know if there is anything else.

stuhood commented on contrib/scrooge/src/python/pants/contrib/scrooge/tasks/scrooge_gen.py:
```diff
@@ -158,6 +171,12 @@ def gen(self, partial_cmd, target, target_workdir):
+      for include_path in partial_cmd.include_paths:
+        args.extend(['--include-path', include_path])
+
+    #raise TaskError('AHA! args are now: {}'.format(str(args)))
```
xx

stuhood commented on src/python/pants/backend/codegen/thrift/java/java_thrift_library.py:
```diff
@@ -39,6 +41,11 @@ def __init__(self,
+      namespace is not explicitly specified in the IDL. The default is defined in the global
+      options under ``--thrift-default-default-java-namespace``.
+    :param include_paths: The include_path(s) used to search for included IDL files. The default
+      is defined in the global options under ``--thrift-default-include-paths``.
```
IMO, let's not document this one. And if I can convince you to remove this parameter entirely, that would be preferable.

I understand that you want to use it for testing, but AFAICT, it should be possible to define dependencies that result in any interesting set of include-paths. As you saw above, include-paths are already passed to scrooge, and I imagine that if you stare at how they are generated for a few minutes, it will be clear how to cause other paths to be searched.

kevinoliver replied:
gotcha. let me take a crack and see if it can be removed.

kevinoliver commented:
@stuhood thanks for the feedback. I've pushed a change but was unable to successfully remove the include_paths option as it would require moving thrift files to new source roots in scrooge.
